webui/react/src/components/UPlot/UPlotChart.tsx

```diff
@@ -1,166 +1,256 @@
+            const maxMax = Math.max(prevMax ?? Number.MIN_SAFE_INTEGER, bboxMax);
+            const minMin = Math.min(prevMin ?? Number.MAX_SAFE_INTEGER, bboxMin);
+
+            /**
+             * here we are cheating a bit by assuming that a zoom is smaller on both ends
+             * this is to get around the issue of calling it a zoom when setData
+             * causes the scale to go from [99.9, 100.1] to [100, 200]
              */
-            if (!scalesRef.current) scalesRef.current = {};
-            if (isZoomed.current) {
-              scalesRef.current[scaleKey] = uPlot.scales[scaleKey];
-            } else {
-              delete scalesRef.current[scaleKey];
-            }
-            if (Object.keys(scalesRef.current).length === 0) scalesRef.current = undefined;
-          } ],
-        },
-        scales: scalesRef.current,
-        width: chartDivRef.current.offsetWidth,
+            const isZoomed = bboxMax < maxMax && bboxMin > minMin;
+
+            boundsRef.current[scaleKey] = {
+              isZoomed,
+              max: maxMax,
+              min: minMin,
+            };
+          },
+        ],
       },
-      options,
-    ) as uPlot.Options;
+      width,
+    },
+    options || {},
+  ) as uPlot.Options;
+
+const UPlotChart: React.FC<Props> = ({
+  data,
+  focusIndex,
+  options,
+  style,
+  noDataMessage,
+}: Props) => {
+  const chartRef = useRef<uPlot>();
+  const chartDivRef = useRef<HTMLDivElement>(null);
+  const boundsRef = useRef<Record<string, ChartBounds>>({});
+  const [ isReady, setIsReady ] = useState(false);
 
-    const plotChart = new uPlot(optionsExtended, normalizedData, chartDivRef.current);
+  const hasData = data && data.length > 1 && (options?.mode === 2 || data?.[0]?.length);
 
+  const extendedOptions = useMemo(
+    () =>
+      getExtendedOptions(
+        options,
+        boundsRef,
+        chartDivRef.current?.offsetWidth,
+        () => setIsReady(true),
+        () => setIsReady(false),
+      ),
+    [ options ],
+  );
+  const previousExtendedOptions = usePrevious(extendedOptions, undefined);
+
+  useEffect(() => {
     return () => {
-      plotChart.destroy();
+      chartRef?.current?.destroy();
       chartRef.current = undefined;
     };
-  }, [ chartDivRef, hasData, normalizedData, options ]);
+  }, []);
 
-  /*
-   * Chart data when data changes.
-   */
   useEffect(() => {
-    if (!chartRef.current || !normalizedData) return;
-    chartRef.current.setData(normalizedData, isZoomed.current);
-  }, [ normalizedData ]);
+    if (!chartDivRef.current) return;
+    if (shouldRecreate(previousExtendedOptions, extendedOptions, chartRef.current)) {
+      /**
+       * TODO: instead of returning true or false,
+       * return a list of actions/payloads to dispatch
+       */
+      chartRef.current?.destroy();
+      chartRef.current = undefined;
 
-  /*
+      const isZoomed = Object.values(boundsRef.current).some((i) => i.isZoomed === true);
```

Comment:
**nit suggestion:** looks like `isZoomed` is used in both `if (shouldRecreate...) else` blocks, you can just defined it once outside of the if/else block.